### DISCUSSION TITLE
Is it safe to use `functools.lru_cache` in FastAPI app?

### DISCUSSION THREAD

**Discussion by partmor** [fastapi/fastapi#8837] [Questions]
In an example in the [documentation](https://fastapi.tiangolo.com/advanced/settings/#creating-the-settings-only-once-with-lru_cache),  a `functools.lru_cache` is used for caching (in memory) an object that is _shared_ across many requests, in order to avoid reading from disk once per user request.

The **question** is: in a more general setting where the cached function called has several arguments and the corresponding endpoint receives many concurrent requests, **is it _recommended_ to use `functools.lru_cache` as proposed within a FastAPI app even though it is not thread-safe?**

```Python
from fastapi import FastAPI
import functools

app = FastAPI()

@functools.lru_cache
def make_slow_query(date: str) -> str:
    data = ...
    return data


@app.get("/data")
def get_data(date: str):
    return {
        "data": make_slow_query(date),
        ...
    }
```

--- 
### Edit

Is it the usage of `Depends` what makes it "valid"? or could it be implemented without `Depends` (as above)?

```Python
from fastapi import FastAPI, Depends
import functools

app = FastAPI()

@functools.lru_cache
def make_slow_query(date: str) -> str:
    data = ...
    return data


@app.get("/data")
def get_data(queried_data: str = Depends(make_slow_query)):
    return {
        "data": queried_data,
        ...
    }
```

Thank you in advance.

**Comment by coder3112** [comment]
Yes, that is absolutely fine.
In fact, the docs recommend it for things like settings. ([Link](https://fastapi.tiangolo.com/advanced/settings/?h=setti#creating-the-settings-only-once-with-lru_cache))

## Example code:
``` python
from functools import lru_cache

from fastapi import Depends, FastAPI

from . import config

app = FastAPI()


@lru_cache()
def get_settings():
    return config.Settings()


@app.get("/info")
async def info(settings: config.Settings = Depends(get_settings)):
    return {
        "app_name": settings.app_name,
        "admin_email": settings.admin_email,
        "items_per_user": settings.items_per_user,
    }
```

**Comment by partmor** [comment]
> Yes, that is absolutely fine.
> In fact, the docs recommend it for things like settings. ([Link](https://fastapi.tiangolo.com/advanced/settings/?h=setti#creating-the-settings-only-once-with-lru_cache))
> 
> ## Example code:
> ```python
> from functools import lru_cache
> 
> from fastapi import Depends, FastAPI
> 
> from . import config
> 
> app = FastAPI()
> 
> 
> @lru_cache()
> def get_settings():
>     return config.Settings()
> 
> 
> @app.get("/info")
> async def info(settings: config.Settings = Depends(get_settings)):
>     return {
>         "app_name": settings.app_name,
>         "admin_email": settings.admin_email,
>         "items_per_user": settings.items_per_user,
>     }
> ```

Hi, thanks. That is precisely the section of the docs I link in my question. I'd like to know how does that hold without falling apart with concurrent calls. Is it `Depends` what's doing the magic here?

**Comment by coder3112** [comment]
Yes, it is absolutely valid code.
The function is just a wrapper which for the same parameters returns the same value without actually doing the function.
As long as your application is able to manage the requests concurrently, it should not be a problem.
#### Quick Note: you can use ```async``` for the route as well. It wont matter.

PS: I ran this code on my system and it worked.

**Comment by partmor** [comment]
> Yes, it is absolutely valid code.
> The function is just a wrapper which for the same parameters returns the same value without actually doing the function.
> As long as your application is able to manage the requests concurrently, it should not be a problem.
> 
> #### Quick Note: you can use `async` for the route as well. It wont matter.
> PS: I ran this code on my system and it worked.

I'm not questioning what `lru_cache` does, I'm very well aware of that, I'm just concerned with its usability in the context of concurrent requests given that it is not thread safe AFAIK. 

It works locally for me too, but I was wondering if it would also work in a scenario of heavy workload. Thanks

**Comment by coder3112** [comment]
> It works locally for me too, but I was wondering if it would also work in a scenario of heavy workload. Thanks

If it works locally(testing with simple benchmarks), it SHOULD work in heavy workload as well.

**Comment by ycd** [comment]
> I'm just concerned with its usability in the context of concurrent requests given that it is not thread safe AFAIK.

I'm not sure that i understood your question correctly, are you asking about threads or asynchronous I/O?

**Comment by partmor** [comment]
I've been looking around and it **seems** that `lru_cache` should be thread-safe at the time of writing. There was an [issue](https://bugs.python.org/issue28969) (dated ~2017) in Python's issue tracker raising the concern of `lru_cache` not being thread-safe, and its status is marked as solved (although browsing Google, this seems still be unnoticed for many developers)

Are you aware of this @ycd ?

> I'm not sure that i understood your question correctly, are you asking about threads or asynchronous I/O?

Good point. I'll split my question then (disclaimer: my knowledge of `asyncio` is **very** limited):

1. Let's assume we have a single worker process: when multiple requests are made to a FastAPI endpoint, is there a chance of _multithreading_ happening, or is it all `asyncio` behind the scenes? (if this makes sense)
2. If it's _all_ `asyncio`, is it safe to use non-thread-safe objects?

Thanks

**Comment by coder3112** [comment]
> Let's assume we have a single worker process: when multiple requests are made to a FastAPI endpoint, is there a chance of multithreading happening, or is it all asyncio behind the scenes? (if this makes sense)

- Well, ```lru_cache```, itself does not use threading. (Does not mean that it is not thread safe).
- Also, WORKERS AND THREADS ARE NOT THE SAME. A single worker can have multiple threads. So *a single worker application can be multithreaded.* Therefore, multithreading will happen.

> If it's all asyncio, is it safe to use non-thread-safe objects?

Of course, use async where its suitable and dont where its not. As long as you manage the pool correctly and keep safety in mind, it  will be safe.

**Comment by partmor** [comment]
> Well, lru_cache, itself does not use threading. (Does not mean that it is not thread safe).

Not sure about what you're trying to conclude here

> Also, WORKERS AND THREADS ARE NOT THE SAME. A single worker can have multiple threads. So a single worker application can be multithreaded. Therefore, multithreading will happen.

We all know the textbook definition. The question I'm asking is if FastAPI will spawn additional threads in abscence of _user-created_ threads or advanced fuctionality like background tasks.

> Of course, use async where its suitable and dont where its not. As long as you manage the pool correctly and keep safety in mind, it will be safe.

That's like saying, _as long as you never fail, you'll never fail_

**Comment by coder3112** [comment]
> We all know the textbook definition. The question I'm asking is if FastAPI will spawn additional threads in abscence of user-created threads or advanced fuctionality like background tasks.

Oh, FastAPI, by default, will not open any threads. Only functions that you write that use multi threading will create a new thread.

> That's like saying, as long as you never fail, you'll never fail
 I realize I was a bit vague, so here is a bit of context: Say, you create a thread function like so:
``` python
start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return f'Done Sleeping...{seconds}'

for _ in range(10):
    t = threading.Thread(target=do_something, args=[1.5])
    t.start()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')
```
#### Here, the time measured wont be correct because you forgot to join the threads. So, it is a bug because you are not managing the threads properly.

The fixed code would be:
``` python
start = time.perf_counter()

def do_something(seconds):
    print(f'Sleeping {seconds} second(s)...')
    time.sleep(seconds)
    return f'Done Sleeping...{seconds}'

threads = []

for _ in range(10):
    t = threading.Thread(target=do_something, args=[1.5])
    t.start()

for thread in threads:
    thread.join()

finish = time.perf_counter()

print(f'Finished in {round(finish-start, 2)} second(s)')
```
If you write the second code, then it wont happen

### *Basically, FastAPI does not affect safety of your app. (wrt threading)*

Your functions do. In this case```lru_cache``` is thread-safe(atleast from what I see on the net.)

Or maybe you could just ensure it was thread safe like so:
``` python
import threading
from collections import defaultdict
from functools import lru_cache, _make_key

def threadsafe_lru(func):
    func = lru_cache()(func)
    lock_dict = defaultdict(threading.Lock)

    def _thread_lru(*args, **kwargs):
        key = _make_key(args, kwargs, typed=False)  
        with lock_dict[key]:
            return func(*args, **kwargs)

    return _thread_lru
```
Taken from https://noamkremen.github.io/a-simple-threadsafe-caching-decorator.html

**Comment by ycd** [comment]
> Oh, FastAPI, by default, will not open any threads

This is not right, if your function is not a coroutine, FastAPI will run that request inside a threadpool. 

> If it's all asyncio, is it safe to use non-thread-safe objects?

asyncio is not safe for direct usage of `lru_cache`, you may want to use something like [`aiocache`](https://github.com/aio-libs/aiocache) for it.

> Let's assume we have a single worker process: when multiple requests are made to a FastAPI endpoint, is there a chance of multithreading happening, or is it all asyncio behind the scenes? (if this makes sense). 

Yes it is possible, if your function is not a coroutine (`async def`), there is multi-threading happening, i explained quite deeply how it works in here -> https://github.com/tiangolo/fastapi/issues/2776#issuecomment-776659392

**Comment by partmor** [comment]
Thanks @ycd !!

It is clear now. 

I'm thinking of opening a PR to modify, or at least just add a warning to the [section](https://fastapi.tiangolo.com/advanced/settings/#creating-the-settings-only-once-with-lru_cache) of the docs that suggests the usage of `lru_cache`. 

Maybe that example is _simple enough_ to never fail, but it might encourage users (like myself) to use it elsewhere in the application.

Do you think this makes sense?

**Comment by coder3112** [comment]
> This is not right, if your function is not a coroutine, FastAPI will run that request inside a threadpool.

I know that, I meant additional threads which are not related to a co-routine will not be opened unless specified explicitly.
Just very bad at explaining. :(

**Comment by ycd** [comment]
> Maybe that example is simple enough to never fail, but it might encourage users (like myself) to use it elsewhere in the application.

You are right, that example would never fail since it is on the global state, it is registered inside the stack of your main process with PCB (Process control block),  so even when a context switch occurs, using `lru_cache`' will remain completely safe.

**Comment by Menghongli** [comment]
Thanks @ycd for the comprehensive explaination!

Would you mind elaborate more on what makes `lru_cache` not safe for asyncio?

**Comment by mezhaka** [comment]
@partmor this is not an answer to your question, but rather some additional trivia for that topic. I found your question thread trying to understand the issue that I experience:
```python
import functools
import fastapi
import pydantic
import time

class Blah(pydantic.BaseModel):
    foo: str

@functools.cache
def blah_provider():
    blah = Blah(foo="foo")
    time.sleep(0.5)  # let's pretend that Blah construction takes a lot of time
    logger.debug(f"in provider {id(blah)=}")

router = fastapi.APIRouter()

@router.get
def get_blah(blah: Annotated[blah, fastapi.Depends(blah_provider)]):
    return blah
```
I start three simultaneous requests and I manage to get different ids. So it prints:
```
id(blah)=4469771600
id(blah)=4466157648
id(blah)=4469504336
```
I expected the same ID, since the provider is caching the return value. The `cache` documentation, however [says](https://docs.python.org/3/library/functools.html#functools.lru_cache):
> The cache is threadsafe so that the wrapped function can be used in multiple threads. This means that the underlying data structure will remain coherent during concurrent updates.
It is possible for the wrapped function to be called more than once if another thread makes an additional call before the initial call has been completed and cached.
